deltachat-ios/Helper/GiveBackMyFirstResponder.swift

```diff
@@ -24,80 +25,75 @@ extension UIViewController {
-            if #unavailable(iOS 18) {
-                return present(imagePicker as UIViewController, animated: animated, completion: completion)
-            }
+        if #unavailable(iOS 17) { // pre iOS 17
+            imagePicker.returnFirstRespondersOnDismiss()
+        } else if #available(iOS 18, *) { // iOS 18 and up
+            imagePicker.returnFirstRespondersOnDismiss()
         }
-
-        let vc = GiveBackMyFirstResponder(culprit: imagePicker)
-        present(vc, animated: animated, completion: completion)
+        present(imagePicker as UIViewController, animated: animated, completion: completion)
     }
 }
 
-private class GiveBackMyFirstResponder<VC: UIViewController>: FirstResponderReturningViewController {
-    @MainActor var culprit: VC
-
-    @MainActor init(culprit: VC) {
-        self.culprit = culprit
-        super.init(nibName: nil, bundle: nil)
+private var lastRespondersKey: UInt8 = 0
+extension UIViewController {
+    /// Resigns first responders when called and returns first responders when this view is dismissed. Call before presenting.
+    fileprivate func returnFirstRespondersOnDismiss() {
+        Self.swizzleOnce()
+        while let next = UIResponder.currentFirstResponder, next.resignFirstResponder() {
+            self.lastResponders.append(next)
+        }
     }
 
-    @MainActor required init?(coder: NSCoder) {
-        fatalError("init(coder:) has not been implemented")
+    fileprivate var lastResponders: [UIResponder] {
+        get { objc_getAssociatedObject(self, &lastRespondersKey) as? [UIResponder] ?? [] }
+        /// Filter for self because that could create a reference cycle
+        set { objc_setAssociatedObject(self, &lastRespondersKey, newValue.filter { $0 !== self }, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)}
     }
 
-    override func viewDidLoad() {
-        super.viewDidLoad()
-        addChild(culprit)
-        culprit.view.frame = view.frame
-        view.addSubview(culprit.view)
-        culprit.didMove(toParent: self)
+    private var selfOrParentIsBeingDismissed: Bool {
+        parent?.isBeingDismissed ?? isBeingDismissed || isBeingDismissed
     }
-}
 
-private class FirstResponderReturningViewController: UIViewController {
-    private var lastResponders: [UIResponder] = {
-        var lastResponders: [UIResponder] = []
-        while let next = UIResponder.currentFirstResponder, next.resignFirstResponder() {
-            lastResponders.append(next)
-        }
-        return lastResponders
-    }()
-
-    override var isBeingDismissed: Bool {
-        parent?.isBeingDismissed ?? super.isBeingDismissed || super.isBeingDismissed
-    }
-
-    override func viewWillDisappear(_ animated: Bool) {
-        super.viewWillDisappear(animated)
-        if isBeingDismissed, !lastResponders.isEmpty, let newFirstResponder = UIResponder.currentFirstResponder {
+    @objc fileprivate func _viewWillDisappear(_ animated: Bool) {
+        _viewWillDisappear(animated)
```

Comment:
i do not really understand what is happening here. at a first glance, this looks to me as a recursive call, but it is not and is working as expected.

but what is then called here?

Reply:
That is how swizzled methods call the method they replaced (so like super. in override methods), I agree it looks like a recursive call.

Reply:
ahh, now i got it!

if any line of code in the project would deserve a comment, this is the line :)

Reply:
It's because if I would try to call viewWillAppear it would end up here again because this method replaces viewWillAppear through swizzling, so they needed a different mechanism for calling the original method.